### PR TITLE
Cf/clean models

### PR DESCRIPTION
**Pull request by CarolinaFurtado**
# Small changes to the code, including extra verifications and new functions
 
 ## Verifications: 
 
### train: if no dataset exists, gives error
image

### train: if no pretrained model exists, gives error 
image

### train thresholds: if no model exists, gives error
image

### train thresholds: if no datasets exists, gives error
image

### plots ok
image
image

### right model selected - yes
no image but it selects the right model depending on the model name 

#### train
image

#### train_thresholds
image

#### test
image

### CODE REVIEW

File: local_utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import logging
+import GPUtil
+import sys
+import tensorflow as tf
+import os
+os.environ['SM_FRAMEWORK'] = 'tf.keras'
+import segmentation_models
+
+
+def folder_has_files(directory, id):
```
[the remark this answers is not in the document]
Comment (rak5216) [reply local_utils.py:17]: i think it's better named as a counterpart to `remote_folder_exists` in gcp_utils, so perhaps we name it `local_folder_exists` here?

[the remark this answers is not in the document]
Comment (rak5216) [reply local_utils.py:17]: unless the directory in question might actually exist while being empty, and thus checking for files really is the key, but i don't think the local directory can exist AND be empty

**Review by rak5216 — CHANGES_REQUESTED** [review]
this is a really awesome start @CarolinaFurtado ! just some cosmetic changes really

**Comment by CarolinaFurtado** [comment]
Awesome! I'll look over and make the changes tomorrow! Thanks!

File: local_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CarolinaFurtado) [reply local_utils.py:17]: It does create a folder with no files, that is why I'm checking if there are files inside the folder. Because it makes a folder, and them copies stuff there (and maybe copies nothing):

```
def copy_folder_locally_if_missing(folder_remote_source, local_folder_dir):
    if not os.path.exists(local_folder_dir.as_posix()):
        local_folder_dir.mkdir(parents=True, exist_ok=True)
        os.system("gsutil -m cp -r '{}' '{}'".format(folder_remote_source, local_folder_dir.as_posix()))
```

[the remark this answers is not in the document]
Comment (CarolinaFurtado) [reply local_utils.py:17]: I'm going to go with `local_folder_has_files` for the name

[the remark this answers is not in the document]
Comment (rak5216) [reply local_utils.py]: awesome, thx for the explanation!

File: models.py [hunk 2]
```diff
@@ -16,15 +16,43 @@
...
-from segmentation_models import Linknet
+from segmentation_models import Unet, FPN, Linknet
 from segmentation_models.losses import CategoricalCELoss
 
 
 thresholds_training_history = {}
 train_thresholds_counter = 0
 
+# These are the only optimizer currently supported
+optimzer_dict = {'adam': Adam(),
```
Comment (rak5216) [inline models.py:27]: sry to be picky, but forgot an 'i':
`optimizer_dict`

File: models.py [hunk 3]
```diff
@@ -149,25 +151,13 @@ def generate_compiled_segmentation_model(model_name, model_parameters, num_class
...
-        if model_parameters['backbone_name'] in FPN_backbones:
-            model = FPN(classes=num_classes, **model_parameters)
-        else:
-            raise NameError("Error, model and backbone are not compatible.")
-    elif model_name == "Linknet":
-        if model_parameters['backbone_name'] in Linknet_backbones:
-            model = Linknet(classes=num_classes, **model_parameters)
-        else:
-            raise NameError("Error, model and backbone are not compatible.")
+
```
Comment (rak5216) [inline models.py:154]: add high-level if-else to make sure `[model_name]` is in `models_dict`, proceed if true but throw helpful error message if choosing a currently unsupported option

File: models.py [hunk 4]
```diff
@@ -37,42 +65,16 @@ def generate_compiled_segmentation_model(model_name, model_parameters, num_class
...
-        opt_fn = Nadam()
-    elif optimizer.lower() == 'rmsprop':
-        opt_fn = RMSprop()
-    else:
-        raise NameError("Optimizer not supported")
+    # Select the optimizer as a function of the name in the config file
+    optimizer_fn = optimzer_dict[optimizer.lower()]
+
+    # Select the loss function  as a function of the name in the config file
+    loss_fn = loss_dict[loss.lower()]
```
Comment (rak5216) [inline models.py:72]: add if-else to make sure `[loss.lower()]` is in `loss_dict`, proceed to assignment if true but throw helpful error message if choosing a currently unsupported option

File: models.py [hunk 5]
```diff
@@ -37,42 +65,16 @@ def generate_compiled_segmentation_model(model_name, model_parameters, num_class
...
-    elif optimizer.lower() == 'ftrl':
-        opt_fn = Ftrl()
-    elif optimizer.lower() == 'nadam':
-        opt_fn = Nadam()
-    elif optimizer.lower() == 'rmsprop':
-        opt_fn = RMSprop()
-    else:
-        raise NameError("Optimizer not supported")
+    # Select the optimizer as a function of the name in the config file
+    optimizer_fn = optimzer_dict[optimizer.lower()]
```
Comment (rak5216) [inline models.py:69]: add if-else to make sure `[optimizer.lower()]` is in `optimizer_dict`, proceed to assignment if true but throw helpful error message if choosing a currently unsupported option

**Review by rak5216 — CHANGES_REQUESTED** [review]

**Review by rak5216 — APPROVED** [review]
